src/main/java/org/variantsync/diffdetective/variationtree/VariationTreeNode.java

```diff
@@ -0,0 +1,364 @@
+package org.variantsync.diffdetective.variationtree;
+
+import org.prop4j.Node;
+import org.variantsync.diffdetective.diff.DiffLineNumber;
+import org.variantsync.diffdetective.diff.difftree.DiffNode; // For Javdoc
+import org.variantsync.diffdetective.diff.difftree.DiffTree; // For Javdoc
+import org.variantsync.diffdetective.diff.difftree.DiffType;
+import org.variantsync.diffdetective.diff.difftree.NodeType;
+import org.variantsync.diffdetective.util.Assert;
+import org.variantsync.diffdetective.util.fide.FixTrueFalse;
+
+import java.util.*;
+
+/**
+ * A single node in a variation tree.
+ *
+ * <p>A variation tree is a representation of source code in a software product line. Each node in
+ * such a tree is either an {@link isArtifact artifact} or a {@link isAnnotation mapping node}.
+ * Artifacts represent source code which may be reused in multiple variations. In contrast, mappings
+ * do not contain source code of a specific variant. They store information about which variants
+ * contain the artifacts it has as {@link getChildren children} in the form of a
+ * {@link getFeatureMapping feature mapping}.
+ *
+ * <p>See definition 2.2 of <a href="https://doi.org/10.1145/3540250.3549108">our ESEC/FSE'22
+ * paper</a> for a mathematical formalization or variation trees. In contrast to this formalization,
+ * this implementation optimizes for variation trees obtained from source code annotated with the C
+ * preprocessor syntax. This mostly means that there is the additional {@link NodeType#ELIF} node
+ * type.
+ *
+ * <p>This class contains references to all of its children and its parent so all connected nodes of
+ * a variation tree can be reached through each node of the variation tree. Nevertheless, most of
+ * the time only node itself or its subtree (the reflexive hull of {@link getChildren}) is meant
+ * when when referencing a {@code VariationTreeNode}. Use {@link VariationTree} to unambiguously
+ * refer to a whole variation tree.
+ *
+ * <p>If possible, algorithms should be using {@link VariationNode} instead of this concrete
+ * implementation. This allows the usage of a projection of a {@link DiffNode} instead of a concrete
+ * variation node.
+ *
+ * <p>To compare two variation trees use {@link DiffTree} which uses the aforementioned
+ * {@link DiffNode}s.
+ *
+ * @see VariationTree
+ * @see DiffTree
+ * @see DiffNode
+ * @author Benjamin Moosherr
+ */
+public class VariationTreeNode extends VariationNode<VariationTreeNode> {
+    private static final short ID_OFFSET = 3;
+
+    /**
+     * The node type of this node, which determines the type of the represented
+     * element in the diff (e.g., mapping or artifact).
+     */
+    public final NodeType nodeType;
+
+    /**
+     * The start line number of the {@link label} of this node in the corresponding source code.
+     */
+    private int from = DiffLineNumber.InvalidLineNumber;
```

Comment:
Explicit naming? `fromInclusive`? Reusing `Lines` here would probably cause more trouble than it helps avoiding ambiguities with offsets.

Reply:
Well, actually ditching the `getFromLine` and `getAfterLine` methods in favor of only keeping one `getLineRange` method seems pretty nice. It keeps me from writing all such verbose methods and attributes and concentrates it into `Lines`.
For review I implement that proposal and add it as a separate commit. Just tell me if you don't like it.

Reply:
See this commit: eb2134152e184d7bd2c245a0f132141e2f2d5eec

Reply:
I like it. We only have to be careful with the variables now being references to objects and no primitive values anymore. I had some bugs with that in the past but it seems to be as if you have been careful here, by making the set-methods side-effect free in terms of returning a new value.

Reply:
Both `Lines` and `DiffLineNumber` are immutable to this should no longer be a problem.

Reply:
Yes! :)